src/nupic/types/Sdr.cpp

```diff
@@ -173,9 +173,18 @@ namespace nupic {
     }
 
 
-    void SparseDistributedRepresentation::setSparse( SDR_sparse_t &value ) {
+    void SparseDistributedRepresentation::setSparse(SDR_sparse_t &value ) {
+	#ifdef NTA_ASSERTIONS_ON
+          const auto orig = SDR_sparse_t(value);
+        #endif
         sparse_.swap( value );
         setSparseInplace();
+        #ifdef NTA_ASSERTIONS_ON
+	  NTA_ASSERT(orig.size() == value.size()) << "orig " << orig.size() << " vs. " << value.size();
```

Comment:
@ctrl-z-9000-times this is a behavior I discovered, and that I think should be fixed: 
setSparse() (probably others too) make a non-const input vector empty. 

```
const vector input1{...};
SDR.setSparse(input1);
//input1 remains the same, as expected :+1: 

vector input2; //not const
SDR.setSparse(input2);
//input2 is cleared! 
```

I suspect it's the `swap` operation, and that the changed input would not have to be empty, but actually replaced by whatever was in `sparse_` (as we "swap"). 

adding const to the method's param signature should fix the problem, 

> void SparseDistributedRepresentation::setSparse(const SDR_sparse_t &value )

Reply:
> adding const to the method's param signature should fix the problem, 

not helping. 
the `swap` is the cause. So, is this known and advertized behavior?

Reply:
Yes, it's the swap operation causing this.  If you cast the input argument to `const` then it should force the SDR to use a different method overload which copies instead of swapping.  Unfortunately both the constant overload & all of the documentation are in the header file.
```
    /**
     * Swap a new value into the SDR, replacing the current value.  This
     * method is fast since it copies no data.  This method modifies its
     * argument!
     *
     * @param value A sparse vector<UInt> to swap into the SDR.
     */
    void setSparse( SDR_sparse_t &value );

    /**
     * Copy a vector of sparse indices of true values.  These indicies are into
     * the flattened SDR space.  This overwrites the SDR's current value.
     *
     * @param value A vector of flat indices to copy into the SDR.
     */
    template<typename T>
    void setSparse( const vector<T> &value ) {
      sparse_.assign( value.begin(), value.end() );
      setSparseInplace();
    }
```

Reply:
> Unfortunately both the constant overload & all of the documentation are in the header file. 

ok, my bad. So it's a documented behavior, I'll ignore that "error"